Data/Aeson/Types/Internal.hs

```diff
@@ -62,25 +66,32 @@ import Data.Vector (Vector)
 import qualified Data.HashMap.Strict as H
 import qualified Data.Vector as V
 
+data JsonPathStep = Key Text | Index Int deriving (Eq, Show)
```

Comment:
Since you're exposing these new types as part of the public API, they need documentation.

Also, please spread out the declaration of `JsonPathStep` across multiple lines (see `Result` for an indication of how I like to format this).

Finally, `JsonPathStep` should probably be named `Segment`.